File: firstLastPosOfElement.java
//Time: O(log n)
//Space: O(1)

public class firstLastPosOfElement {
    int first;
    private int binarySearchFirst(int[] nums, int target){
        int low = 0;
        int high = nums.length - 1;
        
        while(low <= high){
            int mid = low + (high - low)/2;

            if(nums[mid] == target) {
                if (mid == 0 || nums[mid - 1] < nums[mid]){
                    return mid;
                }
                else{
                    high = mid - 1;
                }
            }
            else if (nums[mid] > target){
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }

        }
        return -1;
    }
    private int binarySearchLast(int[] nums, int target){
        int low = first;
        int high = nums.length - 1;
        
        while(low <= high){
            int mid = low + (high - low)/2;

            if(nums[mid] == target) {
                if (mid == nums.length - 1 || nums[mid + 1] > nums[mid]){
                    return mid;
                }
                else {
                    low = mid + 1;
                }
            }
            else if (nums[mid] > target){
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }
        return -1;
    }

    public int[] searchRange(int[] nums, int target) {
        //null case
        if(nums == null || nums.length == 0) return new int[] {-1,-1};
        if(target < nums[0] || target > nums[nums.length - 1]) return new int[] {-1,-1};

        first = binarySearchFirst(nums, target);
        if(first == -1) return new int[] {-1,-1};
        int lastIns = binarySearchLast(nums, target);

        return new int[] {first, lastIns};

    }
}

File: first_last_pos_of_element.py
#!/usr/bin/env python
# coding: utf-8

# In[ ]:


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        
        if (len(nums) == 0 or nums == None): return [-1, -1]
        if target < nums[0]:
            return [-1, -1]
        if (target > nums[len(nums)-1]): return [-1, -1]
        
        firstOccur = self.BinarySearchLeft(nums, target)
        if (firstOccur == -1): return [-1, -1]
        lastOccurence = self.BinarySearchRight(nums, target, firstOccur)
        return [firstOccur, lastOccurence]
        
    def BinarySearchLeft(self, nums, target):
        low = 0
        high = len(nums) - 1
        while (low <= high):
            mid = low + (high - low ) // 2
            if nums[mid] == target:
                if (mid == 0 or nums[mid-1] < nums[mid]):
                    return mid
                high = mid -1
            elif (target < nums[mid]):
                high = mid -1
            else:
                low = mid +1 
        return -1
        
    def BinarySearchRight(self, nums: List[int], target: int, firstOccurence):
        low = firstOccurence
        high = len(nums) -1
        while (low <= high):
            mid = low + (high - low ) // 2
            if nums[mid] == target:
                if (mid == len(nums) - 1 or nums[mid+1] > nums[mid]):
                    return mid
                low = mid + 1
            elif (target < nums[mid]):
                high = mid -1
            else:
                low = mid +1 
        return -1

